packages/liquidator/src/liquidationStrategy.js

```diff
@@ -0,0 +1,223 @@
+    if (shouldLiquidate({ tokensToLiquidate })) {
+      // We should liquidate the entire position using our capital - wdf
+      return createLiquidationParams({
+        sponsor: position.sponsor,
+        tokensToLiquidate,
+        currentBlockTime,
+        maxCollateralPerToken
+      });
+    }
+    // we dont liquidate, dont return anything. Probably because amount to liquidate is 0
+  }
+
+  // Pure function which just calculates how far a timer has progressed between 0 and 100
+  // duration - total duration of the liveness in seconds
+  // end - end time that withdraw passes in unix seconds ts
+  // now - current block time in unix seconds ts
+  function withdrawProgressPercent(duration, end, now) {
+    const elapsed = end - now;
+    if (elapsed <= 0) return 100;
+    if (elapsed >= duration) return 0;
+    return 100 * (1 - elapsed / duration);
+  }
+
+  // Should we try to delay this withdrawal. Update this logic to change conditions to run delay strategy.
+  function shouldLiquidateMinimum({ position, syntheticTokenBalance, empMinSponsorSize, currentBlockTime }) {
+    assert(defenseActivationPercent >= 0, "requires defenseActivationPercent");
+    syntheticTokenBalance = toBN(syntheticTokenBalance);
+    empMinSponsorSize = toBN(empMinSponsorSize);
+
+    // this position does not have a withdraw pending
+    if (position.withdrawalRequestPassTimestamp == 0) return false;
+    // withdraw has passed liveness
+    if (position.withdrawalRequestPassTimestamp <= currentBlockTime) return false;
+    // our synthetic balance is less than the amount required to extend deadline
+    if (syntheticTokenBalance.lt(empMinSponsorSize)) return false;
+    // we have enough to fully liquidate position, so do not do the minimum
+    if (syntheticTokenBalance.gte(toBN(position.numTokens))) return false;
+    // position cant go below the minimum emp sponsor size
+    if (
+      toBN(position.numTokens)
+        .sub(empMinSponsorSize)
+        .lt(empMinSponsorSize)
+    )
+      return false;
+    // liveness timer on withdraw has not passed time
+    if (
+      withdrawProgressPercent(withdrawalLiveness, position.withdrawalRequestPassTimestamp, currentBlockTime) <
+      parseFloat(defenseActivationPercent)
+    )
+      return false;
+
+    // all conditions passed and we should minimally liquidate to extend timer
+    return true;
+  }
+
+  // Any new constraints can be added here, but for now
+  // only dont liquidate if the amount to liquidate is <=0
+  function shouldLiquidate({ tokensToLiquidate }) {
+    tokensToLiquidate = toBN(tokensToLiquidate);
+    // nothing to liquidate
+    if (tokensToLiquidate.lte(toBN(0))) return false;
+    // any other conditions we do try to liquidate
+    return true;
+  }
+
+  return {
+    shouldLiquidate,
+    shouldLiquidateMinimum,
+    createLiquidationParams,
+    processPosition,
+    withdrawProgressPercent,
+    calculateTokensToLiquidate
+  };
```

Comment:
How many of these are public and private? It might be nice to standardize a naming convention for private member functions vs public ones because importing this "class" may be difficult to reason about without some guide as to what function you should pay attention to.

Reply:
We've generally preceded private functions with "_"'s (underscore).

Reply:
In this case, we could even do something more extreme like not exporting the private functions at all (although this may make testing difficult) or putting the private functions inside a sub-object called `private`. However, on the margin, it seems more consistent to just use underscores since that's a normal convention and what we do everywhere else.

Reply:
Yes i need to get better at documenting things here.  I think because this is an "internal" library, the risk of exposing private functions is really low, but it still needs to be documented. If this was a public library, only public functions would be exposed to avoid this issue. I'm not really a fan of renaming functions with "_", since the idea here is you could compose these functions independently for a higher level use, and the "_" implies you should not call them externally, which i dont necessarily agree with.  In this particular case only `processPosition` is used externally, though the rest do get unit tested, so I would prefer to just leave a comment separating private from public.

Reply:
Not super strongly opinionated here, but I think being thoughtful about what the _api surface_ is (even with an internal library) really helps readability. For instance, if I were to see a call to the `processPosition` function (or wanted to use it), I would probably follow it to understand what's going on. If I jumped into this library, it would be tough for me to tell what the API surface is. On first glance, it looks very complex, as if I need to understand 6 functions to really get how to use it. So I think encapsulating complexity by making internal and external functionality very obvious -- done right, I think the user should immediately realize they just need to understand how to use processPosition to _get_ this library and all the other complexity is just in service of that function.

Reply:
Ok I am leaning towards nesting these additional functions under some key like, 'utils'. In this case they are sort of like utlilities, and normally would break them into a seperate file but they share some of the closure state.  so bascially what this looks like is
```js
return {
  // the main api call
  processPosition,
  // utility calls
  utils:{
    shouldLiquidate,
    ...etc,
  }
}
```